### PR TITLE
Repository Editor

### PR DESCRIPTION
**Pull request by zmrow**
Best to read these commits in order:

* `Add from_path() to Target` : This just adds a method to `Target` so we can create one from a `Path`

* `Move RootDigest to tough`: This moves the `RootDigest` abstraction out of `tuftool` into `tough`

* `Add RepositoryEditor`: This is where I need the most help.

### CODE REVIEW

File: tough/src/editor.rs [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+#[derive(Debug)]
+pub struct RepositoryEditor {
+    pub snapshot: Snapshot,
+    pub timestamp: Timestamp,
+    pub targets: Targets,
+    targets_meta: TargetMapTransport,
+}
+
+impl RepositoryEditor {
+    pub fn new(snapshot: Snapshot, timestamp: Timestamp, targets: Targets) -> RepositoryEditor {
```
Comment (webern) [inline tough/src/editor.rs:37]: I don't know if this is a valid concern or not, but the only thing that seems strange to me is that you are taking the three known Role types as input. Maybe it should take the Repo itself as input? In the future there may be additional role types (https://github.com/awslabs/tough/issues/5).

And once the above observation occurs, then I wonder why these operations aren't mutable functions on a Repository itself?

File: tough/src/editor.rs [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+            path.to_owned(),
+        );
+        self.targets_meta
+            .targets
+            .insert(target_name, path.to_owned());
+
+        Ok(())
+    }
+
+    pub fn build<'a, P1, P2, T>(
```
Comment (webern) [inline tough/src/editor.rs:84]: I would probably break this into a couple of subroutines. For example where you have the comments and separators, those seem like good subroutine delineations for refactor.

File: tough/src/editor.rs [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+            timestamp: signed_timestamp,
+            targets: signed_targets,
+            targets_base_url: Url::parse("target_map:///").context(error::ParseUrl {
+                url: "target_map:///",
+            })?,
+        };
+        Ok(repo)
+    }
+
+    fn sign_role<T, P>(
```
Comment (webern) [inline tough/src/editor.rs:209]: Maybe this argument storm could become a struct, like SigningData or something.

**Review by webern — APPROVED** [review]
Seems legit. 👨‍🎨

**Comment by zmrow** [comment]
We've decided to change course a bit and my strategy will change accordingly... closing.

**Pull request closed without merging**